index.js

```diff
@@ -14,255 +12,246 @@ var timedOut = require('timed-out');
 var prependHttp = require('prepend-http');
 var lowercaseKeys = require('lowercase-keys');
 var isRedirect = require('is-redirect');
-var NestedErrorStacks = require('nested-error-stacks');
+var pinkiePromise = require('pinkie-promise');
+var unzipResponse = require('unzip-response');
+var createErrorClass = require('create-error-class');
 
-function GotError(message, nested) {
-	NestedErrorStacks.call(this, message, nested);
-	objectAssign(this, nested, {nested: this.nested});
-}
-
-util.inherits(GotError, NestedErrorStacks);
-GotError.prototype.name = 'GotError';
-
-function got(url, opts, cb) {
-	if (typeof url !== 'string' && typeof url !== 'object') {
-		throw new GotError('Parameter `url` must be a string or object, not ' + typeof url);
-	}
+function requestAsEventEmitter(opts) {
+	opts = opts || {};
 
-	if (typeof opts === 'function') {
-		cb = opts;
-		opts = {};
-	}
+	var ee = new EventEmitter();
+	var redirectCount = 0;
 
-	opts = objectAssign(
-		{
-			protocol: 'http:'
-		},
-		typeof url === 'string' ? urlLib.parse(prependHttp(url)) : url,
-		opts
-	);
+	function get(opts) {
+		var fn = opts.protocol === 'https:' ? https : http;
+		var url = urlLib.format(opts);
 
-	opts.headers = objectAssign({
-		'user-agent': 'https://github.com/sindresorhus/got',
-		'accept-encoding': 'gzip,deflate'
-	}, lowercaseKeys(opts.headers));
+		var req = fn.request(opts, function (res) {
+			var statusCode = res.statusCode;
+			if (isRedirect(statusCode) && 'location' in res.headers && (opts.method === 'GET' || opts.method === 'HEAD')) {
+				res.resume();
 
-	if (opts.pathname) {
-		opts.path = opts.pathname;
-	}
+				if (++redirectCount > 10) {
+					ee.emit('error', new got.MaxRedirectsError({statusCode: statusCode, url: url}), null, res);
+					return;
+				}
 
-	if (opts.query) {
-		if (typeof opts.query !== 'string') {
-			opts.query = querystring.stringify(opts.query);
-		}
+				var redirectUrl = urlLib.resolve(url, res.headers.location);
+				var redirectOpts = objectAssign({}, opts, urlLib.parse(redirectUrl));
 
-		opts.path = opts.pathname + '?' + opts.query;
-		delete opts.query;
-	}
+				ee.emit('redirect', res, redirectOpts);
 
-	var encoding = opts.encoding;
-	var body = opts.body;
-	var json = opts.json;
-	var timeout = opts.timeout;
-	var proxy;
-	var redirectCount = 0;
+				get(redirectOpts);
+				return;
+			}
 
-	delete opts.encoding;
-	delete opts.body;
-	delete opts.json;
-	delete opts.timeout;
+			res = unzipResponse(res);
 
-	if (json) {
-		opts.headers.accept = opts.headers.accept || 'application/json';
-	}
+			ee.emit('response', res);
+		}).once('error', function (err) {
+			ee.emit('error', new got.RequestError(err.message, {
+				url: url,
+				method: opts.method,
+				code: err.code
```

Comment:
`MaxRedirectsError` uses `statusCode`, while this uses `code`. We should make it consistent.